### PR TITLE
#17361: Fix remainder low pcc

### PR DESCRIPTION
**Pull request by mcw-anasuya**
### Ticket
#17361

### Problem description
ttnn.remainder unary low PCC when scalar is between -0.003 and 0.003

### What's changed
As per [comment](https://github.com/tenstorrent/tt-metal/issues/17361#issuecomment-2691371528), to increase the overall accuracy of ttnn.remainder
- floor operation has been reworked in remainder kernel
- `value` and `recip` are loaded as fp32
- Updated golden function for bf16 since torch.bfloat16 was giving inaccurate torch results for tensor-scalar case

### Increase in Overall Accuracy
I tested 10 scalar values in the range [-100, 100] with original logic and with current logic (in this PR). Here are the results:
**Input tensor shape:** [1, 1, 32, 32]
**Input tensor range:** [-100, 100]
**Scalar range:** [-100, 100]
Screenshot 2025-03-28 at 15 57 01

### Observation
Inconsistency in Torch and TT outputs when the input is a multiple of the scalar value (in which case result must be 0).
**Examples in Torch:** 
torch.remainder(torch.tensor([3]), 1.5) = 0.0
torch.remainder(torch.tensor([3]), 0.003) = 0.003
**Examples in TT:**
ttnn.remainder([3], 0.003) = 0.0
ttnn.remainder([4], 0.004) = 0.004

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/14127030333) - PASSED
- [x] [Blackhole Post commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/14123091510) - PASSED
- [ ] [(Single-card) Tests for new models](https://github.com/tenstorrent/tt-metal/actions/runs/13923359778) 
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h
Comment (KalaivaniMCW) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:1]: were you able to test the changes in BH machine ? was there any difference in behaviour ?

File: tests/sweep_framework/sweeps/eltwise/binary/remainder/remainder_forge.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 
 parameters = {
-    "xfail": {
+    "nightly": {
         "input_shape": [[1]],
         "input_a_dtype": [ttnn.int32],
```
Comment (KalaivaniMCW) [inline tests/sweep_framework/sweeps/eltwise/binary/remainder/remainder_forge.py:25]: int32 works fine with vFloat LLKs ?

Comment (umadevimcw) [reply tests/sweep_framework/sweeps/eltwise/binary/remainder/remainder_forge.py:25]: Do we have int32 support?

**Review by umadevimcw — APPROVED** [review]

**Comment by umadevimcw** [comment]
Please test the PR in BH as well

**Review by VirdhatchaniKN — APPROVED** [review]

**Review by umadevimcw — APPROVED** [review]

File: ttnn/ttnn/operations/binary.py [hunk 2]
```diff
@@ -333,13 +333,22 @@ def _golden_function_floor_div(input_tensor_a, input_tensor_b, *args, **kwargs):
 def _golden_function_remainder(input_tensor_a, input_tensor_b, *args, device, **kwargs):
     import torch
 
-    return torch.nan_to_num(
+    input_dtype = input_tensor_a.dtype
+    if not torch.is_tensor(input_tensor_b):
```
[the remark this answers is not in the document]
Comment (mcw-anasuya) [reply ttnn/ttnn/operations/binary.py:337]: Unary remainder computation with torch.bfloat16 has lesser accuracy than ttnn.bfloat16 since we are using fp32 in kernel. 
```
scalar = 0.004
tt input: ttnn.Tensor([[ 5, 2], [3, 4]], shape=Shape([2, 2]), dtype=DataType: :BFLOAT16, layout=Layout:: TILE)
Torch out: tensor ([[0.0027, 0.0011],
                    [0.0016, 0.0022]], dtype=torch.bfloat16)
TT Out: TorchTensor ([[0.0040, 0.0040],
                      [0.0040, 0.0040]], dtype=torch.bfloat16)
```
When I used float() on torch input, the outputs matched.

File: tests/sweep_framework/sweeps/eltwise/binary/remainder/remainder_forge.py (hunk 1 above)
Comment (mcw-anasuya) [reply tests/sweep_framework/sweeps/eltwise/binary/remainder/remainder_forge.py]: The test passes for int32 for the shapes I have tested. For now, I have changed it back to `xfail` suite. We may have to revisit to check how int32 is being supported.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h
Comment (mcw-anasuya) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:1]: Yes, all the test cases pass in BH

**Review by KalaivaniMCW — APPROVED** [review]
Looks good ! We need to document the kernel changes done to improve precision.

**Review by patrickroberts — APPROVED** [review]

**Review by dchenTT — APPROVED** [review]

**Review by mouliraj-mcw — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_remainder.py [hunk 3]
```diff
@@ -47,3 +47,44 @@ def test_broken_remainder1(input_shapes, device):
...
     result = ttnn.to_torch(tt_result)  # all 0.5
     assert torch.allclose(result, golden, atol=0.01, rtol=0)
+
+
+# This test was added for #17361
+# If input is a multiple of the scalar, the result should be 0, but both Torch and TT output either 0 or the scalar value itself depending on the operands.
+# This inconsistency is persistent due to some fp precision loss in both Torch and TT.
+# Eg: torch.remainder of (3, 1.5) = 0.0 and of (3, 0.003) = 0.003
+# Eg: ttnn.remainder of (4, 0.004) = 0.004 and of (3, 0.003) = 0.0
+@skip_for_grayskull("Op not supported for Grayskull, supported for wormhole_b0")
```
[the remark this answers is not in the document]
Comment (mcw-anasuya) [reply tests/ttnn/unit_tests/operations/eltwise/test_remainder.py]: It isn't required for APC, but I'm not sure if any other pipelines test on Grayskull.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h [hunk 4]
```diff
@@ -14,31 +14,43 @@ using namespace sfpi;
 namespace ckernel {
 namespace sfpu {
 
+template <bool APPROXIMATION_MODE>
+inline void _init_remainder_(const uint value, const uint recip) {
```
Comment (rdjogoTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:18]: No need for the `_*_` naming for the init function, since this is in the tt-metal repo and not in the LLK submodule. Don't want to start mixing them up... `init_remainder` is just fine.

**Review by rdjogoTT — APPROVED** [review]
looks good, just change init function naming

File: tests/ttnn/unit_tests/operations/eltwise/test_remainder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ayerofieiev-tt) [reply tests/ttnn/unit_tests/operations/eltwise/test_remainder.py]: grayskull is completely deprecated now. No cards on CI. I recommend to remove those whenever you see them.

[the remark this answers is not in the document]
Comment (mcw-anasuya) [reply tests/ttnn/unit_tests/operations/eltwise/test_remainder.py]: @ayerofieiev-tt Sure, will do that. I've removed it for this test.

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp [hunk 5]
```diff
@@ -141,7 +141,10 @@ std::pair<std::string, std::string> get_op_init_and_func_parameterized(
             break;
         case UnaryOpType::REMAINDER:
             op_init_and_name = {
-                "remainder_tile_init();",
+                fmt::format(
+                    "remainder_tile_init({:#x}u, {:#x}u);",
+                    std::bit_cast<uint32_t>(param0),
+                    std::bit_cast<uint32_t>(1.0f / param0)),
```
Comment (philei-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp:147]: Can we just change llk function to take floats, and don't do bit_cast here to simplify code?

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h [hunk 6]
```diff
@@ -14,31 +14,43 @@ using namespace sfpi;
...
+template <bool APPROXIMATION_MODE>
+inline void init_remainder(const uint value, const uint recip) {
+    // load vConstFloatPrgm0 = value
+    _sfpu_load_config32_(0xC, (value >> 16) & 0xFFFF, value & 0xFFFF);
+    // load vConstFloatPrgm1 = recip
+    _sfpu_load_config32_(0xD, (recip >> 16) & 0xFFFF, recip & 0xFFFF);
+}
+
 template <bool APPROXIMATION_MODE, int ITERATIONS = 8>
 inline void calculate_remainder(const uint value, const uint recip) {
```
Comment (philei-tt) [inline tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:26]: Did you benchmark the kernel? What is the impact of your changes on the performance?

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp (hunk 5 above)
Comment (mcw-anasuya) [reply ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp:147]: This change is out of scope for this PR. It will be an infra change on both ttnn and LLK side. We'll have to discuss with LLK team about this.

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h (hunk 6 above)
Comment (mcw-anasuya) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:26]: Since we are loading both values as fp32, there is a slight delay in kernal duration.
**Old Kernel Duration**: 753866 ns
**New Kernel Duration**: 892658 ns
**Loss**: 138792 ns
However, there is significant improvement in the overall accuracy of the op. For instance, the test range provided in this PR has ATOL < 0.001 with the current changes. Without these changes, the ATOL goes up to 90.0 for the same test cases.

Comment (philei-tt) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:26]: 18% if quite significant 🤔 @ayerofieiev-tt @KalaivaniMCW @patrickroberts What do you think? 
Maybe we can try kernel selection based on some heuristics?

**Review by ayerofieiev-tt — APPROVED** [review]

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h (hunk 6 above)
Comment (patrickroberts) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:26]: Isn't this what `APPROXIMATION_MODE` is supposed to be for? `if constexpr (APPROXIMATION_MODE) { /* old algorithm */ } else { /* new algorithm */ }` (assuming `true` means "fast approximation")

Comment (KalaivaniMCW) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:26]: @patrickroberts @philei-tt  ATOL seems high for some values before this precision change, which is why we are loading both values with fp32 precision. This has improved results outside of -0.003 and 0.003 range as well. Is it ok if one mode gives result with such a big difference from the golden ( torch ) ?

Comment (mcw-anasuya) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_remainder.h:26]: @philei-tt @patrickroberts @KalaivaniMCW @ayerofieiev-tt 
I compared the PCC for input range [-100, 100] with randomised scalar values in the range [-100, 100]. This is the increase in accuracy I found:
Screenshot 2025-03-28 at 15 57 01
